tools/clang/unittests/HLSLExec/ExecutionTest.cpp

```diff
@@ -13236,6 +13205,173 @@ void ExecutionTest::WaveSizeTest() {
+              L"which only supports range [%d, %d]",
+              minShaderWaveSize, maxShaderWaveSize, prefShaderWaveSize,
+              minWaveSize, maxWaveSize);
+          continue;
+        }
+
+        // format compiler args
+        char compilerOptions[64];
+        VERIFY_IS_TRUE(
+            sprintf_s(compilerOptions, sizeof(compilerOptions),
+                      "-D MINWAVESIZE=%d -D MAXWAVESIZE=%d -D PREFWAVESIZE=%d",
+                      minShaderWaveSize, maxShaderWaveSize,
+                      prefShaderWaveSize) != -1);
+
+        // run the shader
+        std::shared_ptr<ShaderOpTestResult> test = RunShaderOpTestAfterParse(
+            pDevice, m_support, "WaveSizeRangeTest",
+            [&](LPCSTR Name, std::vector<BYTE> &Data, st::ShaderOp *pShaderOp) {
+              VERIFY_IS_TRUE((0 == strncmp(Name, "UAVBuffer0", 10)));
+              pShaderOp->Shaders.at(0).Arguments = compilerOptions;
+              pShaderOp->Shaders.at(0).Text = waveSizeTestShader;
+
+              VERIFY_IS_TRUE(sizeof(WaveSizeTestData) * MAX_WAVESIZE <=
+                             Data.size());
+              WaveSizeTestData *pInData = (WaveSizeTestData *)Data.data();
+              memset(pInData, 0, sizeof(WaveSizeTestData) * MAX_WAVESIZE);
+            },
+            ShaderOpSet);
+
+        // verify expected values
+        MappedData dataUav;
+        WaveSizeTestData *pOutData;
+
+        LogCommentFmt(
+            L"Verifying test result for these shader wave size ranges:");
+        LogCommentFmt(L"Minimum shader wave size: %d", minShaderWaveSize);
+        LogCommentFmt(L"Maximum shader wave size: %d", maxShaderWaveSize);
+        LogCommentFmt(L"Preferred shader wave size: %d", prefShaderWaveSize);
+
+        // at this point we assume that the waverange size that
+        // the shader specifies is legal.
+        test->Test->GetReadBackData("UAVBuffer0", &dataUav);
+        VERIFY_ARE_EQUAL(sizeof(WaveSizeTestData) * MAX_WAVESIZE,
+                         dataUav.size());
+        pOutData = (WaveSizeTestData *)dataUav.data();
+
+        for (unsigned i = 0; i < MAX_WAVESIZE; i++) {
+          VERIFY_IS_TRUE(pOutData[i].count >= minWaveSize &&
+                         pOutData[i].count <= maxWaveSize);
+
+          if (prefShaderWaveSize >= minWaveSize &&
+              prefShaderWaveSize <= maxWaveSize) {
+
+            VERIFY_ARE_EQUAL(pOutData[i].count, prefShaderWaveSize);
+          }
+        }
+      }
+    }
+  }
+}
+
+void ExecutionTest::WaveSizeTest() {
+  WEX::TestExecution::SetVerifyOutput verifySettings(
+      WEX::TestExecution::VerifyOutputSettings::LogOnlyFailures);
+
+  CComPtr<ID3D12Device> pDevice;
+  if (!CreateDevice(&pDevice, D3D_SHADER_MODEL_6_6)) {
+    return;
+  }
+
+  // Check Wave support
+  if (!DoesDeviceSupportWaveOps(pDevice)) {
+    // Optional feature, so it's correct to not support it if declared as such.
+    WEX::Logging::Log::Comment(L"Device does not support wave operations.");
```

Comment:
At this point, we can't run any tests, so we should explicitly report skip, right?
```cpp
    WEX::Logging::Log::Result(WEX::Logging::TestResults::Skipped);
```